app/services/prisoner_slot_availability.rb

```diff
@@ -0,0 +1,73 @@
+class PrisonerSlotAvailability
+  PRISONER_UNAVAILABLE = 'prisoner_unavailable'.freeze
+
+  def initialize(prison, noms_id, date_of_birth,
+    date_range = Time.zone.today.to_date..28.days.from_now)
+    self.prison        = prison
+    self.noms_id       = noms_id
+    self.date_of_birth = date_of_birth
+    self.start_date    = date_range.min
+    self.end_date      = date_range.max
+    self.api_error     = false
+  end
+
+  def slots
+    if enforce_all_available_slots?
+      return all_slots_available
+    end
+
+    all_slots_available.each do |slot, unavailability_reasons|
+      unless offender_availabilities_dates.delete(slot.to_date)
```

Comment:
I think this might have a bug. If there are 2 slots for the same date, then only one of the slots will be marked as 'prisoner unavailable' ?

I'd prefer if this method didn't have side effects, so that it would calculate the overall slot availability (`all_slots_available `) each time and also not modify the `offender_availabilities` list.